okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java

```diff
@@ -1243,14 +1243,32 @@ private static String canonicalizeHost(String input, int pos, int limit) {
       if (idnDecoded == null) return null;
 
       // Confirm that the decoded result doesn't contain any illegal characters.
-      int length = idnDecoded.length();
-      if (delimiterOffset(idnDecoded, 0, length, "\u0000\t\n\r #%/:?@[\\]") != length) {
+      if (containsInvalidHostnameCharacters(idnDecoded)) {
         return null;
       }
 
       return idnDecoded;
     }
 
+    private static boolean containsInvalidHostnameCharacters(String idnDecoded) {
+      for (int i = 0; i < idnDecoded.length(); i++) {
```

Comment:
Seems unnecessary to switch to unicode codepoints: the input is guaranteed to be ASCII at this point. I've rearranged the code to try to make that clearer.

FYI - delimiterOffset, the call I replaced, still uses char-by-char comparison. I didn't look to see if that's valid.

Happy to change it if you feel passionate about it.

Regarding blacklist Vs whitelist: could do. I was trying to keep it as close as possible to the WHATWG spec.